

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>Decode cell sub-population based on kinetic parameters of pancreatic endocrinogenesis data &mdash; cellDancer  documentation</title>
  

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
        <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    
    <link rel="author" title="About these documents" href="../about.html" />
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="prev" title="Analyze RNA velocity for mouse hippocampal dentate gyrus neurogenesis data" href="Neuro_velocity_analysis.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html" class="icon icon-home"> cellDancer
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption" role="heading"><span class="caption-text">cellDancer</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../about.html">About cellDancer</a></li>
<li class="toctree-l1"><a class="reference internal" href="../API.html">API - cellDancer key applications</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Tutorials</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="Data%20Preparation.html">Data Preparation</a></li>
<li class="toctree-l1"><a class="reference internal" href="CaseStudy_Gastrulation_transcriptional_boost.html">Resolve the transcriptional boost in gastrulation erythroid maturation data</a></li>
<li class="toctree-l1"><a class="reference internal" href="Neuro_velocity_analysis.html">Analyze RNA velocity for mouse hippocampal dentate gyrus neurogenesis data</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Decode cell sub-population based on kinetic parameters of pancreatic endocrinogenesis data</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#Import-packages">Import packages</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Load-the-velocity-estimation-result-of-all-2,000-genes-in-pancreatic-endocrinogenesis-data">Load the velocity estimation result of all 2,000 genes in pancreatic endocrinogenesis data</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Compute-Cell-Velocity">Compute Cell Velocity</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Kinetic-UMAP-based-on-alpha,-beta,-and-gamma">Kinetic UMAP based on alpha, beta, and gamma</a></li>
</ul>
</li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">cellDancer</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>Decode cell sub-population based on kinetic parameters of pancreatic endocrinogenesis data</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="../_sources/notebooks/Pancreas_decoding_kinetics.ipynb.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container div.prompt *,
div.nboutput.container div.prompt *,
div.nbinput.container div.input_area pre,
div.nboutput.container div.output_area pre,
div.nbinput.container div.input_area .highlight,
div.nboutput.container div.output_area .highlight {
    border: none;
    padding: 0;
    margin: 0;
    box-shadow: none;
}

div.nbinput.container > div[class*=highlight],
div.nboutput.container > div[class*=highlight] {
    margin: 0;
}

div.nbinput.container div.prompt *,
div.nboutput.container div.prompt * {
    background: none;
}

div.nboutput.container div.output_area .highlight,
div.nboutput.container div.output_area pre {
    background: unset;
}

div.nboutput.container div.output_area div.highlight {
    color: unset;  /* override Pygments text color */
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    /*background: #f5f5f5;*/
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
    margin: 0;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt .copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
div.rendered_html th {
  font-weight: bold;
}
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}

/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="Decode-cell-sub-population-based-on-kinetic-parameters-of-pancreatic-endocrinogenesis-data">
<h1>Decode cell sub-population based on kinetic parameters of pancreatic endocrinogenesis data<a class="headerlink" href="#Decode-cell-sub-population-based-on-kinetic-parameters-of-pancreatic-endocrinogenesis-data" title="Permalink to this headline">¶</a></h1>
<p>This tutorial shows how to derive cell fates in the embedding space. Moreover, our advantage of decoding cell sub-population based on kinetic parameters will be shown.</p>
<p>In this analysis of the pancreatic endocrinogenesis data, we followed the gene and cell filtering methods of <a class="reference external" href="https://www.nature.com/articles/s41587-020-0591-3">Bergen et.al</a>. 3,696 cells with 2,000 genes are selected.</p>
<div class="section" id="Import-packages">
<h2>Import packages<a class="headerlink" href="#Import-packages" title="Permalink to this headline">¶</a></h2>
<p>To run the notebook locally, <a class="reference external" href="installation.html">Installation</a> could be referred to install the environment and dependencies.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># import packages</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">glob</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">math</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">celldancer</span> <span class="k">as</span> <span class="nn">cd</span>
<span class="kn">import</span> <span class="nn">celldancer.plotting</span> <span class="k">as</span> <span class="nn">cdplt</span>
<span class="kn">from</span> <span class="nn">celldancer.plotting</span> <span class="kn">import</span> <span class="n">colormap</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="Load-the-velocity-estimation-result-of-all-2,000-genes-in-pancreatic-endocrinogenesis-data">
<h2>Load the velocity estimation result of all 2,000 genes in pancreatic endocrinogenesis data<a class="headerlink" href="#Load-the-velocity-estimation-result-of-all-2,000-genes-in-pancreatic-endocrinogenesis-data" title="Permalink to this headline">¶</a></h2>
<p>To display the prediction in the embedding space and to build Uniform Manifold Approximation and Projection (UMAP) of kinetic parameters by using all genes, all the 2,000 genes have already been predicted and could be downloaded from <a class="reference external" href="link">DentateGyrus_cellDancer_estimation.csv</a>. The result was predicted from <a class="reference external" href="link">PancreaticEndocrinogenesis_cell_type_u_s.csv</a> with <code class="docutils literal notranslate"><span class="pre">celldancer.velocity_estimation.velocity(cell_type_u_s,</span> <span class="pre">permutation_ratio=0.5,</span> <span class="pre">n_jobs=8)</span></code>. For the detail of
pre-processing, <a class="reference external" href="data_processing.html">Data Preparation</a> could be referred to.</p>
<p>The prediction result is online at <a class="reference external" href="xxx">PancreaticEndocrinogenesis_cellDancer_estimation.csv</a>. It could be loaded with <code class="docutils literal notranslate"><span class="pre">pandas.read_csv('your_path/PancreaticEndocrinogenesis_cellDancer_estimation.csv')</span></code>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cellDancer_df_path</span> <span class="o">=</span> <span class="s1">&#39;your_path/PancreaticEndocrinogenesis_cellDancer_estimation.csv&#39;</span>
<span class="n">cellDancer_df</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">cellDancer_df_path</span><span class="p">)</span>
<span class="n">cellDancer_df</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>cellIndex</th>
      <th>gene_name</th>
      <th>unsplice</th>
      <th>splice</th>
      <th>unsplice_predict</th>
      <th>splice_predict</th>
      <th>alpha</th>
      <th>beta</th>
      <th>gamma</th>
      <th>loss</th>
      <th>cellID</th>
      <th>clusters</th>
      <th>embedding1</th>
      <th>embedding2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>0</td>
      <td>Scaper</td>
      <td>0.489151</td>
      <td>0.211323</td>
      <td>0.486841</td>
      <td>0.185478</td>
      <td>0.334391</td>
      <td>0.629880</td>
      <td>0.647696</td>
      <td>0.043435</td>
      <td>AAACCTGAGAGGGATA</td>
      <td>Pre-endocrine</td>
      <td>6.143066</td>
      <td>-0.063644</td>
    </tr>
    <tr>
      <th>1</th>
      <td>1</td>
      <td>Scaper</td>
      <td>0.278262</td>
      <td>0.125742</td>
      <td>0.293904</td>
      <td>0.107517</td>
      <td>0.224909</td>
      <td>0.626946</td>
      <td>0.648537</td>
      <td>0.043435</td>
      <td>AAACCTGAGCCTTGAT</td>
      <td>Ductal</td>
      <td>-9.906417</td>
      <td>0.197778</td>
    </tr>
    <tr>
      <th>2</th>
      <td>2</td>
      <td>Scaper</td>
      <td>0.374380</td>
      <td>0.298116</td>
      <td>0.348888</td>
      <td>0.197525</td>
      <td>0.213979</td>
      <td>0.648890</td>
      <td>0.665279</td>
      <td>0.043435</td>
      <td>AAACCTGAGGCAATTA</td>
      <td>Alpha</td>
      <td>7.559791</td>
      <td>0.583762</td>
    </tr>
    <tr>
      <th>3</th>
      <td>3</td>
      <td>Scaper</td>
      <td>0.320497</td>
      <td>0.198031</td>
      <td>0.318165</td>
      <td>0.146120</td>
      <td>0.219611</td>
      <td>0.636208</td>
      <td>0.655523</td>
      <td>0.043435</td>
      <td>AAACCTGCATCATCCC</td>
      <td>Ductal</td>
      <td>-11.283765</td>
      <td>4.218998</td>
    </tr>
    <tr>
      <th>4</th>
      <td>4</td>
      <td>Scaper</td>
      <td>0.239145</td>
      <td>0.141970</td>
      <td>0.251417</td>
      <td>0.106599</td>
      <td>0.192050</td>
      <td>0.631560</td>
      <td>0.652837</td>
      <td>0.043435</td>
      <td>AAACCTGGTAAGTGGC</td>
      <td>Ngn3 high EP</td>
      <td>1.721565</td>
      <td>-4.753407</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>7391995</th>
      <td>3691</td>
      <td>Tmem63a</td>
      <td>0.000000</td>
      <td>0.136270</td>
      <td>0.000447</td>
      <td>0.125414</td>
      <td>0.005472</td>
      <td>0.110095</td>
      <td>0.067564</td>
      <td>0.045492</td>
      <td>TTTGTCAAGTGACATA</td>
      <td>Pre-endocrine</td>
      <td>4.768472</td>
      <td>-1.388353</td>
    </tr>
    <tr>
      <th>7391996</th>
      <td>3692</td>
      <td>Tmem63a</td>
      <td>0.028708</td>
      <td>0.000000</td>
      <td>0.021286</td>
      <td>0.022931</td>
      <td>0.017270</td>
      <td>0.100517</td>
      <td>0.072741</td>
      <td>0.045492</td>
      <td>TTTGTCAAGTGTGGCA</td>
      <td>Ngn3 high EP</td>
      <td>-1.873335</td>
      <td>-4.182650</td>
    </tr>
    <tr>
      <th>7391997</th>
      <td>3693</td>
      <td>Tmem63a</td>
      <td>0.000000</td>
      <td>0.312690</td>
      <td>0.000200</td>
      <td>0.288797</td>
      <td>0.002445</td>
      <td>0.116196</td>
      <td>0.064809</td>
      <td>0.045492</td>
      <td>TTTGTCAGTTGTTTGG</td>
      <td>Ductal</td>
      <td>-9.882250</td>
      <td>-0.105594</td>
    </tr>
    <tr>
      <th>7391998</th>
      <td>3694</td>
      <td>Tmem63a</td>
      <td>0.028052</td>
      <td>0.154164</td>
      <td>0.019519</td>
      <td>0.165354</td>
      <td>0.008143</td>
      <td>0.107124</td>
      <td>0.069812</td>
      <td>0.045492</td>
      <td>TTTGTCATCGAATGCT</td>
      <td>Alpha</td>
      <td>6.612424</td>
      <td>4.531895</td>
    </tr>
    <tr>
      <th>7391999</th>
      <td>3695</td>
      <td>Tmem63a</td>
      <td>0.076389</td>
      <td>0.046049</td>
      <td>0.055438</td>
      <td>0.101639</td>
      <td>0.024521</td>
      <td>0.098170</td>
      <td>0.073695</td>
      <td>0.045492</td>
      <td>TTTGTCATCTGTTTGT</td>
      <td>Epsilon</td>
      <td>3.071044</td>
      <td>1.120432</td>
    </tr>
  </tbody>
</table>
<p>7392000 rows × 14 columns</p>
</div></div>
</div>
</div>
<div class="section" id="Compute-Cell-Velocity">
<h2>Compute Cell Velocity<a class="headerlink" href="#Compute-Cell-Velocity" title="Permalink to this headline">¶</a></h2>
<p>The prediction in the embedding space could be calculated with <code class="docutils literal notranslate"><span class="pre">celldancer.compute_cell_velocity.compute()</span></code>. If not specified, all genes in cellDancer_df will be used to calculate the prediction in the embedding space. Following the method of <a class="reference external" href="https://velocyto.org/velocyto.py/index.html#">velocyto</a> and <a class="reference external" href="https://scvelo.readthedocs.io/">scVelo</a>, we project the velocity vectors of the cells into the low-dimensional embedding space.</p>
<p>After the calculation, the predicted direction in the embedding space, i.e. columns ‘velocity1’ and ‘velocity2’ will be added to the original dataframe. <code class="docutils literal notranslate"><span class="pre">celldancer.plotting.cell.scatter_cell()</span></code> could be used to display the predicted direction in embedding space.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Compute cell velocity</span>
<span class="n">cellDancer_df</span><span class="o">=</span><span class="n">cd</span><span class="o">.</span><span class="n">compute_cell_velocity</span><span class="o">.</span><span class="n">compute</span><span class="p">(</span><span class="n">cellDancer_df</span><span class="o">=</span><span class="n">cellDancer_df</span><span class="p">,</span> <span class="n">projection_neighbor_size</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>

<span class="c1"># Plot cell velocity</span>
<span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span><span class="mi">10</span><span class="p">))</span>
<span class="n">im</span> <span class="o">=</span> <span class="n">cdplt</span><span class="o">.</span><span class="n">cell</span><span class="o">.</span><span class="n">scatter_cell</span><span class="p">(</span><span class="n">ax</span><span class="p">,</span> <span class="n">cellDancer_df</span><span class="p">,</span> <span class="n">colors</span><span class="o">=</span><span class="n">colormap</span><span class="o">.</span><span class="n">colormap_pancreas</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">velocity</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">legend</span><span class="o">=</span><span class="s1">&#39;on&#39;</span><span class="p">,</span> <span class="n">min_mass</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">arrow_grid</span><span class="o">=</span><span class="p">(</span><span class="mi">20</span><span class="p">,</span><span class="mi">20</span><span class="p">))</span>
<span class="n">ax</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;off&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_Pancreas_decoding_kinetics_11_0.png" src="../_images/notebooks_Pancreas_decoding_kinetics_11_0.png" />
</div>
</div>
</div>
<div class="section" id="Kinetic-UMAP-based-on-alpha,-beta,-and-gamma">
<h2>Kinetic UMAP based on alpha, beta, and gamma<a class="headerlink" href="#Kinetic-UMAP-based-on-alpha,-beta,-and-gamma" title="Permalink to this headline">¶</a></h2>
<p>After obtaining the prediction result, <code class="docutils literal notranslate"><span class="pre">celldancer.embedding_kinetic_para.embedding()</span></code> could build UMAP based on the predicted alpha, beta, gamma, or all of the three.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cellDancer_df</span><span class="o">=</span><span class="n">cd</span><span class="o">.</span><span class="n">embedding_kinetic_para</span><span class="o">.</span><span class="n">embedding</span><span class="p">(</span><span class="n">cellDancer_df</span><span class="p">,</span><span class="s1">&#39;alpha&#39;</span><span class="p">)</span>
<span class="n">cellDancer_df</span><span class="o">=</span><span class="n">cd</span><span class="o">.</span><span class="n">embedding_kinetic_para</span><span class="o">.</span><span class="n">embedding</span><span class="p">(</span><span class="n">cellDancer_df</span><span class="p">,</span><span class="s1">&#39;beta&#39;</span><span class="p">)</span>
<span class="n">cellDancer_df</span><span class="o">=</span><span class="n">cd</span><span class="o">.</span><span class="n">embedding_kinetic_para</span><span class="o">.</span><span class="n">embedding</span><span class="p">(</span><span class="n">cellDancer_df</span><span class="p">,</span><span class="s1">&#39;gamma&#39;</span><span class="p">)</span>
<span class="n">cellDancer_df</span><span class="o">=</span><span class="n">cd</span><span class="o">.</span><span class="n">embedding_kinetic_para</span><span class="o">.</span><span class="n">embedding</span><span class="p">(</span><span class="n">cellDancer_df</span><span class="p">,</span><span class="s1">&#39;alpha_beta_gamma&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>UMAP could be visualized with <code class="docutils literal notranslate"><span class="pre">celldancer.plotting.cell.plot_kinetic_para()</span></code> depending on the selection of kinetic parameter(s). This result shows that our kinetic parameters could reflect cell sub-population. The spliced reads of one gene could be visualized when specifying the particular gene in <code class="docutils literal notranslate"><span class="pre">celldancer.plotting.cell.plot_kinetic_para()</span></code>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">ncols</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">16</span><span class="p">,</span><span class="mi">8</span><span class="p">))</span>
<span class="n">cd</span><span class="o">.</span><span class="n">cdplt</span><span class="o">.</span><span class="n">cell</span><span class="o">.</span><span class="n">plot_kinetic_para</span><span class="p">(</span><span class="n">ax</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="s1">&#39;alpha_beta_gamma&#39;</span><span class="p">,</span> <span class="n">cellDancer_df</span><span class="p">,</span> <span class="n">color_map</span><span class="o">=</span><span class="n">colormap</span><span class="o">.</span><span class="n">colormap_pancreas</span><span class="p">)</span>
<span class="n">cd</span><span class="o">.</span><span class="n">cdplt</span><span class="o">.</span><span class="n">cell</span><span class="o">.</span><span class="n">plot_kinetic_para</span><span class="p">(</span><span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="s1">&#39;alpha&#39;</span><span class="p">,</span> <span class="n">cellDancer_df</span><span class="p">,</span> <span class="n">color_map</span><span class="o">=</span><span class="n">colormap</span><span class="o">.</span><span class="n">colormap_pancreas</span><span class="p">)</span>
<span class="n">cd</span><span class="o">.</span><span class="n">cdplt</span><span class="o">.</span><span class="n">cell</span><span class="o">.</span><span class="n">plot_kinetic_para</span><span class="p">(</span><span class="n">ax</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span> <span class="s1">&#39;beta&#39;</span><span class="p">,</span> <span class="n">cellDancer_df</span><span class="p">,</span> <span class="n">color_map</span><span class="o">=</span><span class="n">colormap</span><span class="o">.</span><span class="n">colormap_pancreas</span><span class="p">)</span>
<span class="n">cd</span><span class="o">.</span><span class="n">cdplt</span><span class="o">.</span><span class="n">cell</span><span class="o">.</span><span class="n">plot_kinetic_para</span><span class="p">(</span><span class="n">ax</span><span class="p">[</span><span class="mi">3</span><span class="p">],</span> <span class="s1">&#39;gamma&#39;</span><span class="p">,</span> <span class="n">cellDancer_df</span><span class="p">,</span> <span class="n">color_map</span><span class="o">=</span><span class="n">colormap</span><span class="o">.</span><span class="n">colormap_pancreas</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;matplotlib.collections.PathCollection at 0x7fca1af66f10&gt;
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_Pancreas_decoding_kinetics_16_1.png" src="../_images/notebooks_Pancreas_decoding_kinetics_16_1.png" />
</div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="Neuro_velocity_analysis.html" class="btn btn-neutral float-left" title="Analyze RNA velocity for mouse hippocampal dentate gyrus neurogenesis data" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2022, 1.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>